zstash/globus.py

```diff
@@ -101,343 +129,386 @@ def globus_activate(hpss: str):
+        ACTIVE, SUCCEEDED, FAILED. The script every 20 seconds polls a
+        status of the transfer job (task) from the Globus Transfer service,
+        with 20 second timeout limit. If the task is ACTIVE after time runs
+        out 'task_wait' returns False, and True otherwise.
+        """
+        if not globus_info.transfer_client:
+            raise ValueError("transfer_client is undefined")
+        while not globus_info.transfer_client.task_wait(
+            task_id, timeout=300, polling_interval=20
+        ):
+            pass
+        """
+        The Globus transfer job (task) has been finished (SUCCEEDED or FAILED).
+        Check if the transfer SUCCEEDED or FAILED.
+        """
+        if not globus_info.transfer_client:
+            raise ValueError("transfer_client is undefined")
+        task = globus_info.transfer_client.get_task(task_id)
+        if task["status"] == "SUCCEEDED":
+            src_ep = task["source_endpoint_id"]
+            dst_ep = task["destination_endpoint_id"]
+            label = task["label"]
+            logger.info(
+                f"Globus transfer {task_id}, from {src_ep} to {dst_ep}: {label} succeeded"
+            )
+        else:
+            logger.error("Transfer FAILED")
+    except TransferAPIError as e:
+        if e.code == "NoCredException":
+            logger.error(
+                f"{e.message}. Please go to https://app.globus.org/endpoints and activate the endpoint."
+            )
+        else:
+            logger.error(e)
+        sys.exit(1)
+    except Exception as e:
+        logger.error(f"Exception: {e}")
+        sys.exit(1)
 
-    for entry in archive_directory_listing:
-        if entry.get("name") == name:
-            return True
-    return False
 
+# Primary functions ###########################################################
 
-global_variable_tarfiles_pushed = 0
+
+def globus_activate(globus_info: GlobusInfo, alt_hpss: str = ""):
+    """
+    Read the local globus endpoint UUID from ~/.zstash.ini.
+    If the ini file does not exist, create an ini file with empty values,
+    and try to find the local endpoint UUID based on the FQDN
+    """
+    if alt_hpss != "":
+        globus_info.hpss_path = alt_hpss
+        globus_info.url = urlparse(alt_hpss)
+        if globus_info.url.scheme != "globus":
+            raise ValueError(f"Invalid url.scheme={globus_info.url.scheme}")
+    globus_info.remote_endpoint = globus_info.url.netloc
+    set_local_endpoint(globus_info)
+    if globus_info.remote_endpoint.upper() in HPSS_ENDPOINT_MAP.keys():
+        globus_info.remote_endpoint = HPSS_ENDPOINT_MAP.get(
+            globus_info.remote_endpoint.upper()
+        )
+    log_current_endpoints(globus_info)
+    set_clients(globus_info)
+    # Causes globus_sdk.services.auth.errors.AuthAPIError:
+    # ('POST', 'https://auth.globus.org/v2/oauth2/token', None, 400, 'Error', 'invalid_grant')
+    # check_consents(globus_info)
```

Comment:
Trying to check consents as early as possible (i.e., as soon as we have a `transfer_client`) just results in an `invalid_grant` error, so that's no good.

Reply:
@forsyth2 At some point, we should rope in some of those helpful globus service folk, and ask why we cannot obtain this information independent of issuing a transfer.  This seems like a "useful" feature to me.